### ISSUE TITLE
RC2: SA1300 for enum value `_12key`

### ISSUE THREAD

**Issue by qmfrederik** [DotNetAnalyzers/StyleCopAnalyzers#1935]
I have an enum that includes a value named `_12key`.

This raises SA1300: Element `_12key` must begin with an uppercase letter.

There are a couple of issues with this:
1. The code fix provider suggest to rename the value to `_12key1`, which does not solve the issue
2. Given that the first three characters are not alphabetic, I would expect the violation not to be flagged or to be flagged for the lowercase `k`. However, renaming to `_12Key` does not fix the issue, either

**Comment by pdelvo** [comment]
Hi,
I think we should do the following: Dont report SA1300 and report SA1309 instead(maybe we have to fix the docs for that). SA1309 should not offer a code fix because the enum value can't start with a digit and there is not sensible way to fix it automatically.

**Comment by sharwell** [comment]
The violation is reported based on #1859. The decision to treat enum members as elements which are not constants (i.e. to report SA1300 instead of SA1303) was intentional.

I don't think any changes need to be made to the analyzer as a result of this.

The `RenameToUpperCaseCodeFixProvider` code fix could be improved in a manner similar to the way we already improved the `RenameToLowerCaseCodeFixProvider` in #1861. Before making any changes we should narrow down the exact desired semantics, which I haven't spent too much time on yet.

**Comment by qmfrederik** [comment]
Hi @sharwell 

I'm fine with SA1300 applying to enum values; the issue is that in my opinion, SA1300 should not be reported for an element named `_12clock`.

Here's why 
* Classic StyleCop is fine with it. `_12clock` or just `_12` results in no warning; `clock` results in SA1300, element names must begin with an upper case letter
* I would say starting an element name with `_` is justified if the subsequent character requires you to do so (in this case, element names cannot start with digits, so you must use an underscore)
* Similarly, if your element name starts with digits, I would say the SA1300 rule should not return a warning since the rule does not apply
* You could argue that the text comes after the digits should be capitalized, as you should treat `12clock` as a compound name and Pascal casing requires the individual words of a compound name to be capitalized.

Net, probably `_12Clock` would probably be "correct", but `_12clock` does not violate SA1300 because it starts with digits and I would consider digits to be uppercased. 
This would be consistent with classic StyleCop

**Comment by sharwell** [comment]
@vweijsters What does StyleCop Classic report for the following code?

```csharp
public class _12clock
{
}
```

Also for the following code:

```csharp
public class _12Clock
{
}
```

**Comment by vweijsters** [comment]
Nothing is reported for both cases.

**Comment by vweijsters** [comment]
I would be OK with adding an exception to SA1300 for elements starting with an underscore followed by a number (following @qmfrederik argument that the underscore is justified because the number is otherwise not allowed).
